### PR TITLE
Support pushing a manifest list from a layout

### PR DESCRIPTION
**Pull request by glyn**
Note to reviewer:
* it may be more digestible to read each commit separately
* (if you merge): please retain the three commits in the history, for the same reason as above, with a merge commit to group them together logically.

Fixes https://github.com/pivotal/image-relocation/issues/34

### CODE REVIEW

File: pkg/registry/ggcr/client.go [hunk 1]
```diff
@@ -32,8 +51,12 @@ type client struct {
 	writeRemoteIndex indexWriter
 }
 
+var _ RegistryClient = &client{}
+
+// TODO: *client implements Digest and Copy too
```
Comment (vdice) [inline pkg/registry/ggcr/client.go]: Shall we address this TODO in this PR as well?

**Comment by glyn** [comment]
> All looks reasonable to me, though I don't have enough history with this codebase to offer approval of the larger refactors.

Thanks for taking the time and fair enough.

> One question I had was around the removal of the `nm`/`image.Name` field on the `imageManifest` and `imageIndex` structs. What was the motivation to do so? Can we gather similar context from the existing fields -- and if so, should we add the image name context back into the error messages, when thrown? Or, perhaps this context was deemed unnecessary?

The motivation for removing the image name from those structs was that, when constructing the either of the structs from a digest in an OCI image layout, we would need to scan the image index looking for the image name annotation for the entry with the given digest. This is inefficient but it also requires the annotations to be present which may not be the case for OCI image layouts produced by other tools.

Since the image name is only used in error messages, I made it the (internal to this repo) callers' responsibility to add the image name to the error or, in the case of OCI image layout Push, to add the digest to the error.

The net effect is that the command line user or the caller of the external interface will get errors including either the image name or the digest, depending on the inputs to the current operation.

Note that the external interfaces of this repository are intended to be just `Image` and `Layout` (and the methods used to construct instances of these). The lower level public interfaces are, of course exposed, but aren't intended for external consumption.

File: azure-pipelines.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ pool:
 
 variables:
   GOBIN:  '$(GOPATH)/bin' # Go binaries path
-  GOROOT: '/usr/local/go1.11' # Go installation path
+  GOROOT: '/usr/local/go1.12' # Go installation path
```
[the remark this answers is not in the document]
Comment (glyn) [reply azure-pipelines.yml]: By testing with 1.12, we ensure that the code can be compiled with either 1.12 or 1.13. I suggest we upgrade to 1.13 in due course when it's less likely that someone will want to compile it with 1.12.

File: pkg/registry/ggcr/client.go (hunk 1 above)
Comment (glyn) [reply pkg/registry/ggcr/client.go]: It was really a note to self and I've deleted it now.

File: pkg/registry/ggcr/layout_test.go [hunk 3]
```diff
@@ -126,4 +131,75 @@ var _ = Describe("Layout", func() {
 			})
 		})
 	})
+
+	Describe("Push", func() {
+		const testDigest = "sha256:0000000000000000000000000000000000000000000000000000000000000000"
+		var (
+			digest    image.Digest
+			targetRef image.Name
```
[the remark this answers is not in the document]
Comment (glyn) [reply pkg/registry/ggcr/layout_test.go]: The actual value is irrelevant, but it should indeed have a value to make the test more meaningful. Added.

File: pkg/registry/ggcr/client.go (hunk 1 above)
Comment (glyn) [reply pkg/registry/ggcr/client.go]: That interface looked like dead code and I inadvertently deleted it. https://github.com/pivotal/image-relocation/pull/39 will re-instate it (and ensure it isn't dead so I don't make the same mistake again). Apologies!

**Pull request merged**